include/cpp2util.h

```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
 //  Types
 //
 template< typename C, typename X >
-auto is( X const& x ) -> bool {
+constexpr auto is( X const& x ) -> auto {
     if constexpr (
         std::is_same_v<C, X>
         || std::is_base_of_v<C, X>
     )
     {
-        return true;
+        return std::true_type{};
     }
     else if constexpr (
-        std::is_base_of_v<X, C>
-        || (
-            std::is_polymorphic_v<C>
-            && std::is_polymorphic_v<X>
-            )
+        std::is_polymorphic_v<C>
+        && std::is_polymorphic_v<X>
```

Comment:
Is the `is_base_of` alternative redundant? Or is this fixing the semantics of a certain use case?

Reply:
The issue with `std::is_base_of_v<X, C>` is that it will also match non-polymorphic types for which dynamic_cast is undefined.

see: https://godbolt.org/z/GMvfrGzEh

Reply:
Ah, the following check
```
std::is_polymorphic_v<C>
        && std::is_polymorphic_v<X>
```
is enough to know that we can use `dynamic_cast` and it works for all combinations of polymorphic types.

See: https://godbolt.org/z/fhTbE31Ke

Reply:
I want to let you know that this PR is just the beginning of a series of PRs that correct or improve the `is()` and `as()`. The upcoming PRs (some of which I already requested) will improve/correct some of the cases that are still in this code.

I intend to make these changes digestive.